agent/xds/server.go

```diff
@@ -0,0 +1,418 @@
+	reqStop := int32(0)
+	go func() {
+		for {
+			req, err := stream.Recv()
+			if atomic.LoadInt32(&reqStop) != 0 {
+				return
+			}
+			if err != nil {
+				close(reqCh)
+				return
+			}
+			reqCh <- req
+		}
+	}()
+
+	err := s.process(stream, reqCh)
+	if err != nil {
+		s.logger.Printf("[DEBUG] Error handling ADS stream: %s", err)
+	}
+
+	// prevents writing to a closed channel if send failed on blocked recv
+	atomic.StoreInt32(&reqStop, 1)
+
+	return err
+}
+
+const (
+	stateInit int = iota
+	statePendingAuth
+	stateRunning
+)
+
+func (s *Server) process(stream ADSStream, reqCh <-chan *v2.DiscoveryRequest) error {
+	// xDS requires a unique nonce to correlate response/request pairs
+	var nonce uint64
+
+	// xDS works with versions of configs. Internally we don't have a consistent
+	// version. We could just hash the config since versions don't have to be
+	// ordered as far as I can tell, but it's cheaper just to increment a counter
+	// every time we observe a new config since the upstream proxycfg package only
+	// delivers updates when there are actual changes.
+	var configVersion uint64
+
+	// Loop state
+	var cfgSnap *proxycfg.ConfigSnapshot
+	var req *v2.DiscoveryRequest
+	var ok bool
+	var stateCh <-chan *proxycfg.ConfigSnapshot
+	var watchCancel func()
+	var proxyID string
+
+	// need to run a small state machine to get through initial authentication.
+	var state = stateInit
+
+	// Configure handlers for each type of request
+	handlers := map[string]*xDSType{
+		EndpointType: &xDSType{
+			typeURL:   EndpointType,
+			resources: endpointsFromSnapshot,
+			stream:    stream,
+		},
+		ClusterType: &xDSType{
+			typeURL:   ClusterType,
+			resources: clustersFromSnapshot,
+			stream:    stream,
+		},
+		RouteType: &xDSType{
+			typeURL:   RouteType,
+			resources: routesFromSnapshot,
+			stream:    stream,
+		},
+		ListenerType: &xDSType{
+			typeURL:   ListenerType,
+			resources: listenersFromSnapshot,
+			stream:    stream,
+		},
+	}
+
+	for {
+		select {
+		case <-stream.Context().Done():
+			// TODO: Not sure if we should error here, this presumably happens during
+			// a normal shutdown where client goes away too.
+			return nil
+		case req, ok = <-reqCh:
+			if !ok {
+				return nil
+			}
+			if req.TypeUrl == "" {
+				return status.Errorf(codes.InvalidArgument, "type URL is required for ADS")
+			}
+			if handler, ok := handlers[req.TypeUrl]; ok {
+				handler.Recv(req)
+			}
+		case cfgSnap = <-stateCh:
+			// We got a new config, update the version counter
+			configVersion++
+		}
+
+		// Trigger state machine
+		switch state {
+		case stateInit:
+			if req == nil {
+				// This can't happen (tm) since stateCh is nil until after the first req
+				// is recieved but lets not panic about it.
+				continue
+			}
+			// Start authentication process, we need the proxyID
+			proxyID = req.Node.Id
+
+			// Start watching config for that proxy
+			stateCh, watchCancel = s.cfgMgr.Watch(proxyID)
+			defer watchCancel()
```

Comment:
Did you mean to do a `defer` inside of the wrapped `for` loop? Usually that can backfire.  It's hard to tell from the flow of this state machine if this is ok in context.

Reply:
Yep, comment added.